src/file_io.rs

```diff
@@ -0,0 +1,119 @@
+use std::{fs, io, iter::Iterator};
+
+pub trait FsReadWriter {
+    fn read_dir(&self, path: &str) -> io::Result<fs::ReadDir>;
+}
+
+pub struct Fs {}
+
+impl FsReadWriter for Fs {
+    fn read_dir(&self, path: &str) -> io::Result<fs::ReadDir> {
+        fs::read_dir(path)
+    }
+}
+
+pub struct FilterResult {
+    pub paths: Vec<String>,
+    pub io_errors: Vec<io::Error>,
+}
+
+pub struct FsFilterer<FS: FsReadWriter> {
+    fs_rw: FS,
+}
+
+impl FsFilterer<Fs> {
+    pub fn new() -> FsFilterer<Fs> {
+        FsFilterer { fs_rw: Fs {} }
+    }
+}
+
+impl<FS: FsReadWriter> FsFilterer<FS> {
+    pub fn new_custom(fs_rw: FS) -> FsFilterer<FS> {
+        FsFilterer { fs_rw }
+    }
+
+    pub fn dir(
+        &self,
+        path: &str,
+        file_dir_filter: u8,
+        hidden_filter: u8,
+        extension_filter: Option<&str>,
+    ) -> Result<FilterResult, io::Error> {
+        let mut io_errors: Vec<io::Error> = Vec::new();
+        let paths: Vec<String> = self
+            .fs_rw
+            .read_dir(path)?
+            .filter_map(|entry| -> Option<fs::DirEntry> {
+                match entry {
+                    Ok(entry) => Some(entry),
+                    Err(e) => {
+                        io_errors.push(e);
+                        None
+                    }
+                }
+            })
+            .filter_map(|entry| {
+                let path = entry.path();
+                match (fs::metadata(&path), path.file_stem(), path.extension()) {
+                    (Ok(metadata), Some(file_stem), Some(extension)) => {
+                        let is_file = metadata.is_file() as u8;
+                        let is_dir = (metadata.is_dir() as u8) << 1;
```

Comment:
Is `<<` a shift?  Why?

Reply:
Is bit manipulation really necessary here in general?  Seems like all this could be done more in a more readable way with boolean operations.

Reply:
I ended up resorting to bit manipulation because it ends up being really gross and verbose with booleans. The logic you're representing ends up being weirdly tricky. An entry can be a file or a directory, hidden or not hidden, and can optionally be filtered by extension. And then the params allow you to specify that the entry can be hidden, not hidden, or both, etc. This ended up being the simplest way:
- A file is 0b01, a dir is 0b10; the param you provide is either 0b01, 0b10, or 0b11
- Not hidden is 0b01, hidden is 0b10; param is same as above
- Extension checked separately

Reply:
Fair enough, but I do think it could use a bit of method-level documentation.

Reply:
you're right, will add a method-level comment. in fact, the code in general needs some more commenting.

Reply:
I messed up and added comments to a commit on the other PR that had other stuff on it.

Instead of taking the time to copy comments over one at a time, let's just merge this and take the comments on the other.